enhancements/hide-container-mountpoints.md

```diff
@@ -0,0 +1,339 @@
+
+- A convenience wrapper to enter this well-known mount namespace,
+  `kubensenter` for other tools, administrative and support actions which
+  need access to this namespace.
+  - This will operate identically to `nsenter` except that it defaults to
+    entering this well-known namespace location (if present)
+
+- An update to the debug container's MOTD that mentions about the new need for
+  `kubensenter` in addition to the current chroot instructions so it's clear
+  how the debug shell can gain access to the hidden Kubernetes mountpoints.
+
+- Both the new systemd service and convenience wrapper will be installed
+  as part of Kubelet.
+
+With this proposal in place, both Kubelet and CRI-O create their mounts in
+the new shared (with each other) but private (from systemd) namespace, and
+this feature can be easily enabled/disabled by enabling/disabling a single
+systemd service.
+
+### Risks and Mitigations
+
+The current OpenShift and Kubernetes implementations guarantee 3 things about
+mountpoint visibility:
+1. Mounts originating in the host OS are visible to CRI-O, OCI hooks, Kubelet,
+   and container volumeMounts with `MountPropagation: HostToContainer` (or
+   Bidirectional)
+2. Mounts originating in CRI-O, OCI hooks, Kubelet, and container volumeMounts
+   with `mountPropagation: Bidirectional` are visible to each other and
+   container volumeMounts with `MountPropagation: HostToContainer`
+3. Mounts originating in CRI-O, OCI hooks, Kubelet, and container volumeMounts
+   with `mountPropagation: Bidirectional` are visible to the host OS
+
+The first 2 guarantees are not changed by this proposal:
+1. The new mount namespace uses 'slave' propagation, so any mounts originating
+   in the host OS top-level mount namespace are still propagated down into the
+   new 2nd-level namespace where they are visible to CRIO-O, OCI hooks,
+   Kubelet, and container volumeMounts with `MountPropagation: HostToContainer`
+   (or Bidirectional), just as before.
+2. CRI-O, OCI hooks, Kubelet, and any containers created by CRI-O are all
+   within the same 2nd-level namespace, so any mountpoints created by any of
+   these entities are visible to all others within that same mount namespace.
+   Additionally, any 3rd-level namespaces created below this point will have
+   the same relationship with the 2nd-level namespace that they previously had
+   with the higher-level namespace.
+
+The 3rd guarantee is explicitly removed by this proposal.
+
+This means that:
+- Administrators who have connected to the host OS and want to inspect the
+  mountpoints originating from CRI-O, OCI hooks, Kubelet, or containers will
+  not be able to see them unless they explicitly enter the 2nd-level namespace.
+- Any external or 3rd-party tools which run in the host mount namespace but
+  expect to see mountpoints created by CRI-O, OCI hooks, Kubelet, or containers
+  would need to be changed to enter the specific container namespace in order
+  to see them.
```

Comment:
We can call out that this would break security scanning products. Also, need to check what work we need to do if any in stackrock.

Reply:
Who would be best suited to ask about any potential implications for stackrox?

Reply:
Adding @connorgorman for stackrox.

Reply:
Please see the following lines. I'm don't understand the proposal completely as I'm not sure which paths would start being hidden, but these are the paths that we are currently tracking including tapping into systemd to see certain services
https://github.com/stackrox/stackrox/blob/master/compliance/collection/file/file.go#L36-L92

Reply:
@robbycochran Can you take a look at this enhancement and help determine if it will impact collector?

Reply:
@connorgorman and @robbycochran : Here's a quick run-down of how this proposal might affect you, and when it should not.

- If your code runs inside a container, you shouldn't notice a difference except if you're doing privileged operations to access the host OS's mount namespace.
- If your code runs on the host OS (or asks systemd on the host OS for its perspective), the main difference is that systemd and other host OS processes won't have immediate access to any mountpoints created by kubelet or cri-o, or any containers like CSI providers.  This includes the following general classes of mountpoints:
  - Overlay mounts under /var/lib/containers/storage/ and /run/containers/storage/
  - Namespace bindmounts under /run/*ns/
  - secrets and volumes mounted under /var/lib/kubelet/pods/

If you're in this 2nd group, you can always use 'nsenter' or 'setns()' to jump into the kubernetes-specific mount namespace, and all the mounts will then be visible again.

Reply:
@lack StackRox uses privileged containers to access the host OS directories, including `/var/run/docker.sock`, `/proc`, `/etc`, `/usr/lib`, `/sys`, and `/dev` and the other paths listed in the compliance file above.  However, we don't access any mount points created by kubelet or cri-o. So I don't think this change will impact us.

Reply:
@robbycochran I did see `/proc/mounts` listed in the file.go that @connorgorman posted above, and what you get back from there definitely depends on what mount namespace you're in...

Reply:
@lack Just looked into the code /proc/mounts is only used in one case of the compliance container to check that there is a container partition as a part of the Docker benchmark. I think losing this one case is okay as this benchmark is less relevant moving forward without the deprecation of docker bench